Add replication support to dbatools!

jpomfret opened the pull request:
<!-- Below information IS REQUIRED with every PR -->
## Please read -- recent changes to our repo
On November 10, 2022, [we removed some bloat from our repository (for the second and final time)](https://github.com/dataplat/dbatools/issues/8542). This change requires that all contributors reclone or refork their repo.

PRs from repos that have not been recently reforked or recloned will be closed and @potatoqualitee will cherry-pick your commits and open a new PR with your changes.

 - [X] Please confirm you have the smaller repo (85MB .git directory vs 275MB or 110MB or 185MB .git directory)

## Type of Change
<!-- What type of change does your code introduce -->
 - [ ] Bug fix (non-breaking change, fixes #<!--issue number--> )
 - [X] New feature (non-breaking change, adds functionality, fixes #<!--issue number--> )
 - [X] Breaking change (affects multiple commands or functionality, fixes #<!--issue number--> )
 - [ ] Ran manual Pester test and has passed (`.\tests\manual.pester.ps1`)
 - [ ] Adding code coverage to existing functionality
 - [X] Pester test is included
 - [ ] If new file reference added for test, has is been added to github.com/dataplat/appveyor-lab ?
 - [ ] Unit test is included
 - [X] Documentation
 - [ ] Build system

<!-- Below this line you can erase anything that is not applicable -->
### Purpose
Adding replication support to dbatools!

### Approach
There are many commands added with this PR to cover most of the replication functionality we can add - there is still work to do on these commands, more options and settings that can be added - and more code paths. However, they are all working and at this point I want to create this PR and start moving them towards the dbatools dev branch so we can start getting feedback.

I realise I've included the psd1 in this PR - we can remove it before we merge in - just makes testing this branch easier

### Commands to test - New commands

#### Distribution
- Enable-DbaReplDistributor
- Disable-DbaReplDistributor

#### Publishing
- Enable-DbaReplPublishing
- Disable-DbaReplPublishing

#### Publications
- New-DbaReplPublication
- Remove-DbaReplPublication

#### Articles
- Get-DbaReplArticle
- Get-DbaReplArticleColumn
- Add-DbaReplArticle
- Remove-DbaReplArticle
- New-DbaReplCreationScriptOptions

#### Subscriptions
- New-DbaReplSubscription - works but doesn't output anything - waiting on Get-DbaReplSubscription
- Remove-DbaReplSubscription - once Get-DbaReplSubscription works I want to rework this one to enable piping
- Get-DbaReplSubscription -- this command still needs work - can't currently get the subscription information from the subscriber instance

### Commands to test - Existing commands that have been spruced up
- Get-DbaReplServer
- Get-DbaReplDistributor
- Get-DbaReplPublication

### Commands that still need some love
- Test-DbaRepLatency - haven't got to this one yet, but it doesn't work
- Export-DbaRepServerSetting - this exports server settings - could be expanded to include publications etc

## Testing
The following script takes you through building and destroying replication:

```PowerShell
## based off these two containers
# create a shared network
docker network create localnet

# Expose engines and setup shared path for migrations
docker run -p 2500:1433  --volume shared:/shared:z --name mssql1 --hostname mssql1 --network localnet -d dbatools/sqlinstance
docker run -p 2600:1433 --volume shared:/shared:z --name mssql2 --hostname mssql2 --network localnet -d dbatools/sqlinstance2

# create the repl folder
docker exec mssql1 mkdir /var/opt/mssql/ReplData

# also need these folders for setting up replication
docker exec mssql1 mkdir /shared/data /shared/repldata


# using psdefaultparametervalues to make connections easy
$securePassword = ('dbatools.IO' | ConvertTo-SecureString -AsPlainText -Force)
$credential = New-Object System.Management.Automation.PSCredential('sqladmin', $securePassword)

$PSDefaultParameterValues = @{
    "*:SqlCredential"            = $credential
    "*:DestinationCredential"    = $credential
    "*:DestinationSqlCredential" = $credential
    "*:SourceSqlCredential"      = $credential
    "*:PublisherSqlCredential"   = $credential
}


# enable distribution
Enable-DbaReplDistributor -SqlInstance mssql1

# enable publishing
Enable-DbaReplPublishing -SqlInstance mssql1

# Get the distributor
Get-DbaReplDistributor -SqlInstance mssql1

# add a transactional publication
$pub = @{
    SqlInstance = 'mssql1'
    Database    = 'pubs'
    Name        = 'testPub'
    Type        = 'Transactional'
}
New-DbaReplPublication @pub -verbose

# add a merge publication
$pub = @{
    SqlInstance = 'mssql1'
    Database    = 'pubs'
    Name        = 'mergey'
    Type        = 'Merge'
}
New-DbaReplPublication @pub

# add a snapshot publication
$pub = @{
    SqlInstance = 'mssql1'
    Database    = 'pubs'
    Name        = 'snappy'
    Type        = 'Snapshot'
}
New-DbaReplPublication @pub

# view publications
Get-DbaReplPublication -SqlInstance mssql1

# add an article to each publication
$article = @{
    SqlInstance = 'mssql1'
    Database    = 'pubs'
    Publication = 'testpub'
    Name        = 'publishers'
    Filter      = "city = 'seattle'"
}
Add-DbaReplArticle @article

$article = @{
    SqlInstance = 'mssql1'
    Database    = 'pubs'
    Publication = 'Mergey'
    Name        = 'publishers'
}
Add-DbaReplArticle @article

$article = @{
    SqlInstance = 'mssql1'
    Database    = 'pubs'
    Publication = 'snappy'
    Name        = 'publishers'
}
Add-DbaReplArticle @article

# view articles
Get-DbaReplArticle -SqlInstance mssql1

# add subscriptions
$sub = @{
    SqlInstance               = 'mssql2'
    Database                  = 'pubs'
    PublicationDatabase       = 'pubs'
    PublisherSqlInstance      = 'mssql1'
    PublicationName           = 'testpub'
    Type                      = 'Push'
    SubscriptionSqlCredential = $credential

}
New-DbaReplSubscription @sub -enableexception

$sub = @{
    SqlInstance               = 'mssql2'
    Database                  = 'Mergeypubs'
    PublicationDatabase       = 'pubs'
    PublisherSqlInstance      = 'mssql1'
    PublicationName           = 'Mergey'
    Type                      = 'Push'
    SubscriptionSqlCredential = $credential

}
New-DbaReplSubscription @sub

$sub = @{
    SqlInstance               = 'mssql2'
    Database                  = 'Snappypubs'
    PublicationDatabase       = 'pubs'
    PublisherSqlInstance      = 'mssql1'
    PublicationName           = 'Snappy'
    Type                      = 'Push'
    SubscriptionSqlCredential = $credential
}
New-DbaReplSubscription @sub

# remove subscriptions
$sub = @{
    SqlInstance          = 'mssql2'
    SubscriptionDatabase = 'pubs'
    PublisherSqlInstance = 'mssql1'
    PublicationDatabase  = 'pubs'
    PublicationName      = 'testPub'
}
Remove-DbaReplSubscription @sub

$sub = @{
    SqlInstance          = 'mssql2'
    SubscriptionDatabase = 'Mergeypubs'
    PublisherSqlInstance = 'mssql1'
    PublicationDatabase  = 'pubs'
    PublicationName      = 'Mergey'
}
Remove-DbaReplSubscription @sub

$sub = @{
    SqlInstance          = 'mssql2'
    PublisherSqlInstance = 'mssql1'
    PublicationName      = 'snappy'
    PublicationDatabase  = 'pubs'
    SubscriptionDatabase = 'Snappypubs'
}
Remove-DbaReplSubscription @sub

# remove an article
$article = @{
    SqlInstance = 'mssql1'
    Database    = 'pubs'
    Publication = 'testpub'
    Name        = 'publishers'
}
Remove-DbaReplArticle @article

# remove an article
$article = @{
    SqlInstance = 'mssql1'
    Database    = 'pubs'
    Publication = 'Mergey'
    Name        = 'publishers'
}
Remove-DbaReplArticle @article

# remove an article
$article = @{
    SqlInstance = 'mssql1'
    Database    = 'pubs'
    Publication = 'snappy'
    Name        = 'publishers'
}
Remove-DbaReplArticle @article

## remove publications
$pub = @{
    SqlInstance = 'mssql1'
    Database    = 'pubs'
    Name        = 'Snappy'
}
Remove-DbaReplPublication @pub

$pub = @{
    SqlInstance = 'mssql1'
    Database    = 'pubs'
    Name        = 'TestPub'
}
Remove-DbaReplPublication @pub

$pub = @{
    SqlInstance = 'mssql1'
    Database    = 'pubs'
    Name        = 'Mergey'
}
Remove-DbaReplPublication @pub


# disable publishing - bug in Remove-DbaReplPublication - need to check on whether the database gets disabled for replication when last pub gets dropped
Disable-DbaReplPublishing -SqlInstance mssql1 -force

# disable distribution
Disable-DbaReplDistributor -SqlInstance  mssql1
```

potatoqualitee commented:
koooo!

potatoqualitee commented:
for the integration tests, you can put a note that they are in GH if you havent. looking forward to this!

jpomfret commented:
@potatoqualitee I've added unit tests with a note about github actions for integration testing

jpomfret commented:
I'm still working through a couple of test fixes but this is a massive PR so I'm sure there will be some things that need review and fixes so figured I'd change it from draft and we can start that process :)

wsmelton commented on ReplicationCommands.md:
```diff
@@ -0,0 +1,80 @@
+# Replication Work
```
This should be in an issue so it is kept as history when/if this file is ever deleted.

jpomfret replied:
Totally - opened https://github.com/dataplat/dbatools/issues/8995 and I'll remove this file from the repo - thanks!

wsmelton commented on ReplicationDemo.ps1:
```diff
@@ -0,0 +1,99 @@
+# dbatools 💜 dbatools
```
This should be incorporated into the devcontainer

jpomfret replied:
Yep, good shout - will add a TODO and get that done - might move to an issue so it's outside of this PR but will see how far I get this week.

wsmelton replied:
I'm wondering if maybe we just create a dedicated demo repository for this? Then that repo could have a dedicated devcontainer setup to load into VS code for users.

wsmelton replied:
Make it easier if we do this in a dedicated repo just use docker-compose that way users can see how we are building this out and play with adjusting as they want. I can help convert your docker run commands into docker-compose if you need.

We have [dataplat/dbatools-lab](https://github.com/dataplat/dbatools-lab) for the book, so could make say a `dataplat/dbatools-repl-lab` repo.

wsmelton replied:
Was this file going to be removed/moved to a new repo?

wsmelton commented on public/Add-DbaReplArticle.ps1:
```diff
@@ -0,0 +1,179 @@
+        If this switch is enabled, you will be prompted for confirmation before executing any operations that change state.
+
+    .NOTES
+        Tags: Replication
```
```suggestion
        Tags: repl, Replication
```

Cause most know it as repl more than "replication" 🍷

wsmelton commented on public/Disable-DbaReplDistributor.ps1:
```diff
@@ -0,0 +1,102 @@
+        If this switch is enabled, you will be prompted for confirmation before executing any operations that change state.
+
+    .NOTES
+        Tags: Replication
```
```suggestion
        Tags: repl, Replication
```

wsmelton commented on public/Disable-DbaReplPublishing.ps1:
```diff
@@ -0,0 +1,100 @@
+        If this switch is enabled, you will be prompted for confirmation before executing any operations that change state.
+
+    .NOTES
+        Tags: Replication
```
```suggestion
        Tags: repl, Replication
```

wsmelton commented on public/Enable-DbaReplDistributor.ps1:
```diff
@@ -0,0 +1,92 @@
+        If this switch is enabled, you will be prompted for confirmation before executing any operations that change state.
+
+    .NOTES
+        Tags: Replication
```
```suggestion
        Tags: repl, Replication
```

wsmelton commented on public/Enable-DbaReplPublishing.ps1:
```diff
@@ -0,0 +1,118 @@
+        If this switch is enabled, you will be prompted for confirmation before executing any operations that change state.
+
+    .NOTES
+        Tags: Replication
```
```suggestion
        Tags: repl, Replication
```

wsmelton commented on public/Export-DbaReplServerSetting.ps1:
```diff
@@ -1,4 +1,4 @@
-function Export-DbaRepServerSetting {
+function Export-DbaReplServerSetting {
```
Alias should be added so it is not a breaking change for users.

wsmelton replied:
Tag should be updated in this command as well for `repl, Replication`

wsmelton commented on public/Get-DbaRepPublication.ps1:
```diff
@@ -1,120 +0,0 @@
-function Get-DbaRepPublication {
```
I'd prefer the git history be corrected where we have a rename, and then the change in code. This makes it look like we removed the command.

wsmelton replied:
I believe only steps needed to accomplish this...

1. Copy out the current version of the renamed file (one you have currently)
2. Delete that renamed file
3. Checkout this file (old name) from dev branch to reset
4. Rename the file (e.g., `git mv` or Explorer method)
5. Commit that change (this keeps history intact after the rename)
6. Then add your code version back to the file.

wsmelton commented on public/Get-DbaReplPublication.ps1:
```diff
@@ -0,0 +1,151 @@
+        Using this switch turns this "nice by default" feature off and enables you to catch exceptions with your own try/catch.
+
+    .NOTES
+        Tags: Replication
```
```suggestion
        Tags: repl, Replication
```

wsmelton commented on public/Get-DbaReplArticleColumn.ps1:
```diff
@@ -0,0 +1,122 @@
+        Using this switch turns this "nice by default" feature off and enables you to catch exceptions with your own try/catch.
+
+    .NOTES
+        Tags: Replication
```
```suggestion
        Tags: repl, Replication
```

wsmelton commented on public/Get-DbaReplArticle.ps1:
```diff
@@ -0,0 +1,128 @@
+        Using this switch turns this "nice by default" feature off and enables you to catch exceptions with your own try/catch.
+
+    .NOTES
+        Tags: Replication
```
```suggestion
        Tags: repl, Replication
```

wsmelton commented on public/Get-DbaReplPublisher.ps1:
```diff
@@ -0,0 +1,79 @@
+        Using this switch turns this "nice by default" feature off and enables you to catch exceptions with your own try/catch.
+
+    .NOTES
+        Tags: Replication
```
```suggestion
        Tags: repl, Replication
```

wsmelton commented on public/Get-DbaReplSubscription.ps1:
```diff
@@ -0,0 +1,84 @@
+        Using this switch turns this "nice by default" feature off and enables you to catch exceptions with your own try/catch.
+
+    .NOTES
+        Tags: Replication
```
```suggestion
        Tags: repl, Replication
```

wsmelton commented on public/New-DbaReplCreationScriptOptions.ps1:
```diff
@@ -0,0 +1,80 @@
+            MarkReplicatedForeignKeyConstraintsAsNotForReplication, and Schema
+
+    .NOTES
+        Tags: Replication, Script
```
```suggestion
        Tags: repl, Replication, Script
```

wsmelton commented on public/New-DbaReplPublication.ps1:
```diff
@@ -0,0 +1,200 @@
+        If this switch is enabled, you will be prompted for confirmation before executing any operations that change state.
+
+    .NOTES
+        Tags: Replication
```
```suggestion
        Tags: repl, Replication
```

wsmelton commented on public/New-DbaReplSubscription.ps1:
```diff
@@ -0,0 +1,285 @@
+        If this switch is enabled, you will be prompted for confirmation before executing any operations that change state.
+
+    .NOTES
+        Tags: Replication
```
```suggestion
        Tags: repl, Replication
```

wsmelton commented on public/Remove-DbaReplArticle.ps1:
```diff
@@ -0,0 +1,152 @@
+        If this switch is enabled, you will be prompted for confirmation before executing any operations that change state.
+
+    .NOTES
+        Tags: Replication
```
```suggestion
        Tags: repl, Replication
```

wsmelton commented on public/Remove-DbaReplPublication.ps1:
```diff
@@ -0,0 +1,171 @@
+        If this switch is enabled, you will be prompted for confirmation before executing any operations that change state.
+
+    .NOTES
+        Tags: Replication
```
```suggestion
        Tags: repl, Replication
```

wsmelton commented on public/Remove-DbaReplSubscription.ps1:
```diff
@@ -0,0 +1,149 @@
+        If this switch is enabled, you will be prompted for confirmation before executing any operations that change state.
+
+    .NOTES
+        Tags: Replication
```
```suggestion
        Tags: repl, Replication
```

wsmelton commented on public/Add-DbaReplArticle.ps1:
```diff
@@ -0,0 +1,179 @@
+        The name of the object to add as an article.
+
+    .PARAMETER Filter
+        Horizontal filter for replication, implemented as a where clause, but don't include the word WHERE>
```
Is the `>` needed?

wsmelton replied:
```suggestion
        Sets the where clause used to filter the article horizontally, e.g., DiscontinuedDate IS NULL
```
A bit easier to process (considering I haven't touched replication in a bit) and even reading the doc meaning the WHERE keyword I think will throw the beginners off a bit.
Ref example: https://learn.microsoft.com/en-us/dotnet/api/microsoft.sqlserver.replication.transarticle.filterclause?view=sqlserver-2016#microsoft-sqlserver-replication-transarticle-filterclause

wsmelton commented on public/Add-DbaReplArticle.ps1:
```diff
@@ -0,0 +1,179 @@
+    #>
+    [CmdletBinding(DefaultParameterSetName = "Default", SupportsShouldProcess, ConfirmImpact = 'Medium')]
+    param (
+        [parameter(Mandatory, ValueFromPipeline)]
```
```suggestion
        [Parameter(Mandatory, ValueFromPipeline)]
```

wsmelton commented on public/Add-DbaReplArticle.ps1:
```diff
@@ -0,0 +1,179 @@
+        [DbaInstanceParameter[]]$SqlInstance,
+        [PSCredential]$SqlCredential,
+        [parameter(Mandatory)]
+        [String]$Database,
```
```suggestion
        [string]$Database,
```

wsmelton commented on public/Add-DbaReplArticle.ps1:
```diff
@@ -0,0 +1,179 @@
+        [PSCredential]$SqlCredential,
+        [parameter(Mandatory)]
+        [String]$Database,
+        [parameter(Mandatory)]
```
```suggestion
        [Parameter(Mandatory)]
```

wsmelton commented on public/Add-DbaReplArticle.ps1:
```diff
@@ -0,0 +1,179 @@
+        [parameter(Mandatory)]
+        [String]$Database,
+        [parameter(Mandatory)]
+        [String]$Publication,
```
```suggestion
        [string]$Publication,
```

wsmelton replied:
Are we also going to support piping to this command from say `Get-DbaReplPublication`?

wsmelton commented on public/Add-DbaReplArticle.ps1:
```diff
@@ -0,0 +1,179 @@
+        [String]$Database,
+        [parameter(Mandatory)]
+        [String]$Publication,
+        [String]$Schema = 'dbo',
```
```suggestion
        [string]$Schema = 'dbo',
```

wsmelton commented on public/Add-DbaReplArticle.ps1:
```diff
@@ -0,0 +1,179 @@
+        [parameter(Mandatory)]
+        [String]$Publication,
+        [String]$Schema = 'dbo',
+        [parameter(Mandatory)]
```
```suggestion
        [Parameter(Mandatory)]
```

wsmelton commented on public/Add-DbaReplArticle.ps1:
```diff
@@ -0,0 +1,179 @@
+        [String]$Publication,
+        [String]$Schema = 'dbo',
+        [parameter(Mandatory)]
+        [String]$Name,
```
```suggestion
        [string]$Name,
```

wsmelton commented on public/Add-DbaReplArticle.ps1:
```diff
@@ -0,0 +1,179 @@
+        [String]$Schema = 'dbo',
+        [parameter(Mandatory)]
+        [String]$Name,
+        [String]$Filter,
```
```suggestion
        [string]$Filter,
```

wsmelton commented on public/Add-DbaReplArticle.ps1:
```diff
@@ -0,0 +1,179 @@
+        [String]$Name,
+        [String]$Filter,
+        [PSObject]$CreationScriptOptions,
+        [Switch]$EnableException
```
```suggestion
        [switch]$EnableException
```

wsmelton commented on public/Add-DbaReplArticle.ps1:
```diff
@@ -0,0 +1,179 @@
+
+        foreach ($instance in $SqlInstance) {
+            try {
+                $replServer = Get-DbaReplServer -SqlInstance $instance -SqlCredential $SqlCredential
```
```suggestion
                $replServer = Get-DbaReplServer -SqlInstance $instance -SqlCredential $SqlCredential -EnableException:$EnableException
```

Any call to another public function should have the EnableException passed in so downstream will throw exceptions as well.

jpomfret replied:
Good point, thanks!

wsmelton commented on public/Add-DbaReplArticle.ps1:
```diff
@@ -0,0 +1,179 @@
+            try {
+                if ($PSCmdlet.ShouldProcess($instance, "Adding an article to $Publication")) {
+
+                    $pub = Get-DbaReplPublication -SqlInstance $instance -SqlCredential $SqlCredential -Name $Publication
```
```suggestion
                    $pub = Get-DbaReplPublication -SqlInstance $instance -SqlCredential $SqlCredential -Name $Publication -EnableException:$EnableException
```

wsmelton commented on public/Add-DbaReplArticle.ps1:
```diff
@@ -0,0 +1,179 @@
+            Write-Message -Level Verbose -Message "Adding article $Name to publication $Publication on $instance"
+
+            try {
+                if ($PSCmdlet.ShouldProcess($instance, "Adding an article to $Publication")) {
```
Having this be at the top of the workflow does not provide much detail on WhatIf output. Adding this if block to each logical check you are doing with data from the Get command call will provide more details to the user. This by nature will also improve the verbose output of the command too.

jpomfret replied:
Updated and this is the output - thought?

```PowerShell
Add-DbaReplArticle @article -whatif
What if: Performing the operation "Get the publication details for testpub" on target "mssql1".
What if: Performing the operation "Create an article object for testpub which is a Transactional publication" on target "mssql1".
What if: Performing the operation "Add filter for article: publishers" on target "mssql1".
What if: Performing the operation "Create article: publishers" on target "mssql1".
```

wsmelton commented on public/Add-DbaReplArticle.ps1:
```diff
@@ -0,0 +1,179 @@
+            } catch {
+                Stop-Function -Message "Unable to add article $Name to $Publication on $instance" -ErrorRecord $_ -Target $instance -Continue
+            }
+            Get-DbaReplArticle -SqlInstance $instance -SqlCredential $SqlCredential -Publication $Publication -Name $Name
```
```suggestion
            Get-DbaReplArticle -SqlInstance $instance -SqlCredential $SqlCredential -Publication $Publication -Name $Name -EnableException:$EnableException
```

wsmelton commented on public/Disable-DbaReplDistributor.ps1:
```diff
@@ -0,0 +1,102 @@
+    #>
+    [CmdletBinding(DefaultParameterSetName = "Default", SupportsShouldProcess, ConfirmImpact = 'High')]
+    param (
+        [parameter(Mandatory, ValueFromPipeline)]
```
```suggestion
        [Parameter(Mandatory, ValueFromPipeline)]
```

wsmelton commented on public/Disable-DbaReplDistributor.ps1:
```diff
@@ -0,0 +1,102 @@
+    process {
+        foreach ($instance in $SqlInstance) {
+            try {
+                $replServer = Get-DbaReplServer -SqlInstance $instance -SqlCredential $SqlCredential
```
```suggestion
                $replServer = Get-DbaReplServer -SqlInstance $instance -SqlCredential $SqlCredential -EnableException:$EnableException
```

wsmelton commented on public/Disable-DbaReplDistributor.ps1:
```diff
@@ -0,0 +1,102 @@
+                try {
+                    if ($PSCmdlet.ShouldProcess($instance, "Disabling distribution on $instance")) {
+                        # remove any connections to the distribution database
+                        Get-DbaProcess -SqlInstance $instance -SqlCredential $SqlCredential -Database $replServer.DistributionDatabases.name | Stop-DbaProcess
```
```suggestion
                        Get-DbaProcess -SqlInstance $instance -SqlCredential $SqlCredential -Database $replServer.DistributionDatabases.name -EnableException:$EnableException | Stop-DbaProcess -EnableException:$EnableException
```

wsmelton commented on public/Disable-DbaReplDistributor.ps1:
```diff
@@ -0,0 +1,102 @@
+        regardless of whether or not dependent publishing and distribution objects are uninstalled.
+
+    #>
+    [CmdletBinding(DefaultParameterSetName = "Default", SupportsShouldProcess, ConfirmImpact = 'High')]
```
```suggestion
    [CmdletBinding(SupportsShouldProcess, ConfirmImpact = 'High')]
```

wsmelton commented on public/Add-DbaReplArticle.ps1:
```diff
@@ -0,0 +1,179 @@
+        Adds the stores table to the testPub publication from mssql1.pubs with the NonClusteredIndexes and Statistics options set
+        includes default options.
+    #>
+    [CmdletBinding(DefaultParameterSetName = "Default", SupportsShouldProcess, ConfirmImpact = 'Medium')]
```
```suggestion
    [CmdletBinding(SupportsShouldProcess, ConfirmImpact = 'Medium')]
```

wsmelton reviewed (commented):
Have to say reviewing this in the extension in VS Code is way easier now than using the web UI 🧇

wsmelton commented on public/Add-DbaReplArticle.ps1:
```diff
@@ -0,0 +1,179 @@
+function Add-DbaReplArticle {
+    <#
+    .SYNOPSIS
+        Adds an article to a publication for the database on the target SQL instances.
```
```suggestion
        Add an article to a publication of a database on the target SQL Server instance.
```
As the command only works with one article at a time.

wsmelton replied:
```suggestion
        Add an article to a publication of a database on the SQL Server instance.
```
As the command only works with one article at a time.

wsmelton commented on public/Add-DbaReplArticle.ps1:
```diff
@@ -0,0 +1,179 @@
+        Adds an article to a publication for the database on the target SQL instances.
+
+    .DESCRIPTION
+        Adds an article to a publication for the database on the target SQL instances.
```
```suggestion
        Add an article configuration to a publication in a database on the target SQL Server instance.
```

wsmelton commented on public/Add-DbaReplArticle.ps1:
```diff
@@ -0,0 +1,179 @@
+        The database on the publisher that contains the article to be replicated.
+
+    .PARAMETER Publication
+        The name of the replication publication.
```
```suggestion
        The name of the publication.
```

wsmelton commented on public/Add-DbaReplArticle.ps1:
```diff
@@ -0,0 +1,179 @@
+        Adds an article to a publication for the database on the target SQL instances.
+
+    .PARAMETER SqlInstance
+        The target SQL Server instance or instances.
```
This needs to be the publisher instance correct? Can't be the subscriber?


```suggestion
        The SQL Server instance(s) for the publication.
```

wsmelton commented on public/Add-DbaReplArticle.ps1:
```diff
@@ -0,0 +1,179 @@
+        For MFA support, please use Connect-DbaInstance.
+
+    .PARAMETER Database
+        The database on the publisher that contains the article to be replicated.
```
```suggestion
        The publication database to apply the article configuration to be replicated.
```

wsmelton commented on public/Add-DbaReplArticle.ps1:
```diff
@@ -0,0 +1,179 @@
+        The name of the replication publication.
+
+    .PARAMETER Schema
+        The schema name that contains the object to add as an article.
```
```suggestion
        Schema where the article to be added is found.
```

wsmelton commented on public/Add-DbaReplArticle.ps1:
```diff
@@ -0,0 +1,179 @@
+        PS C:\> Add-DbaReplArticle -SqlInstance mssql1 -Database Northwind -Publication PubFromPosh -Name TableToRepl
+
+        Adds the TableToRepl table to the PubFromPosh publication from mssql1.Northwind
+
```
rm extra line

wsmelton commented on public/Add-DbaReplArticle.ps1:
```diff
@@ -0,0 +1,179 @@
+
+                    if ($Filter) {
+                        if ($Filter -like 'WHERE*') {
+                            Stop-Function -Message "Filter should not include the word 'WHERE'" -ErrorRecord $_ -Target $instance -Continue
```
I'd almost prefer this to be a parameter validation so it is caught sooner than having them execute the command 🤔

jpomfret replied:
Makes sense - moved to the top

wsmelton commented on public/Add-DbaReplArticle.ps1:
```diff
@@ -0,0 +1,179 @@
+                    }
+
+                    if (-not ($article.IsExistingObject)) {
+                        $article.Create()
```
It would be cleaner if the try/catch was only focused on this line instead of the whole code block (130-173)

wsmelton commented on public/Add-DbaReplArticle.ps1:
```diff
@@ -0,0 +1,179 @@
+
+                    # need to refresh subscriptions so they know about new articles
+                    # only on trans\snap publications
+                    # (Method invocation failed because [Microsoft.SqlServer.Replication.MergePublication] does not contain a method named 'RefreshSubscriptions'.)
```
Is this comment meant for further details, research, catching?

wsmelton commented on bin/Replication/Invoke-ReplicationSetup.ps1:
```diff
@@ -0,0 +1,13 @@
+
+
+
+Invoke-DbaQuery -File .\bin\Replication\setup-test-replication.sql
```
This file should be added to dataplat/appveyor-lab and referenced from that repository. You can include checking this repo out in the GH action.

jpomfret replied:
I actually ended up removing the need for this - so I'll remove the file from the repo! Sorry missed this one will remove it and the commented out references

wsmelton commented on bin/Replication/setup-test-replication.sql:
```diff
@@ -0,0 +1,40 @@
+-- Set up distribution
```
This file should be added to dataplat/appveyor-lab and referenced from that repository. You can include checking this repo out in the GH action.

jpomfret replied:
No longer required - will remove file.

jpomfret reviewed (commented):
First bit of reviewing the review.. more to come.

Thanks @wsmelton for going through this!

wsmelton commented on public/Disable-DbaReplDistributor.ps1:
```diff
@@ -0,0 +1,102 @@
+function Disable-DbaReplDistributor {
+    <#
+    .SYNOPSIS
+        Disables replication distribution for the target SQL instances.
```
```suggestion
        Disables the Distributor feature for replication on the specified SQL Server instance(s).
```

wsmelton replied:
@jpomfret question on this one because documentation is stating for the method used that it actually uninstalls all replication publishing and distribution on the current instance. This is not "disabling", should we call this `Uninstall-DbaReplication` instead?

Ref: https://learn.microsoft.com/en-us/dotnet/api/microsoft.sqlserver.replication.replicationserver.uninstalldistributor

wsmelton commented on public/Disable-DbaReplDistributor.ps1:
```diff
@@ -0,0 +1,102 @@
+function Disable-DbaReplDistributor {
+    <#
+    .SYNOPSIS
+        Disables replication distribution for the target SQL instances.
```
If you read the remarks that contradicts what the top of the page says 😆

jpomfret replied:
Alright @wsmelton - had a debate with my work team on this and we went in a full circle and ended up back on enable\disable but adding to the confirm messaging `Disabling and removing publishing on $instance"` to hopefully make it more clearer.

Main reasons were enable\disable is what you do in SSMS when you are working with replication, and although we agreed it feels like if you disable something and then enable it you'd get the same thing back, people that work with replication should understand the concepts and how this works.

Install\Uninstall - felt like actually installing the feature through sqlsetup
New\Remove - New makes me think I can do something more than once, but you can only setup publishing\distribution once per instance
Register\Unregister - this was our second choice - but feels a lot like enable\disable and so favoured the wording that matched the SSMS experience.

jpomfret replied:
Yeah makes sense - will wait for @potatoqualitee steer on this one.

wsmelton commented on public/Disable-DbaReplDistributor.ps1:
```diff
@@ -0,0 +1,102 @@
+        Disables replication distribution for the target SQL instances.
+
+    .DESCRIPTION
+        Disables replication distribution for the target SQL instances.
```
```suggestion
        Uninstalls the distributor database on the specific SQL Server instance.
        This command is equivalent to running sp_dropdistributor stored procedure in SQL Server.
```

If this is true as what I linked above and the bottom of that doc page references this as being the equivalent I'd still say this needs to use Uninstall verb.

wsmelton commented on public/Disable-DbaReplDistributor.ps1:
```diff
@@ -0,0 +1,102 @@
+        Boolean value that specifies whether or not replication objects are removed from the server,
+        even if a remote Distributor cannot be reached.
+
+        If true, the publishing and Distributor configuration at the current server is uninstalled regardless of
```
Please remove line breaks

wsmelton commented on public/Disable-DbaReplDistributor.ps1:
```diff
@@ -0,0 +1,102 @@
+        For MFA support, please use Connect-DbaInstance.
+
+    .PARAMETER Force
+        Boolean value that specifies whether or not replication objects are removed from the server,
```
Please remove the line break

wsmelton commented on public/Disable-DbaReplDistributor.ps1:
```diff
@@ -0,0 +1,102 @@
+        If true, the publishing and Distributor configuration at the current server is uninstalled regardless of
+        whether or not dependent publishing and distribution objects are uninstalled.
+
+        If false, the publisher and distribution databases must already be uninstalled, and no local databases
```
Please remove the line break

wsmelton commented on public/Disable-DbaReplDistributor.ps1:
```diff
@@ -0,0 +1,102 @@
+    .EXAMPLE
+        PS C:\> Disable-DbaReplDistributor -SqlInstance mssql1
+
+        Disables replication distribution for the mssql1 instance.
```
Update to Uninstall if we are switching the verb around

wsmelton commented on public/Disable-DbaReplDistributor.ps1:
```diff
@@ -0,0 +1,102 @@
+
+        Disables replication distribution for the mssql1 and mssql2 instances using a sql login.
+        Specifies force so the publishing and Distributor configuration at the current server is uninstalled
+        regardless of whether or not dependent publishing and distribution objects are uninstalled.
```
Please remove line breaks.

Update to Uninstall if we are changing the verb

wsmelton commented on public/Disable-DbaReplDistributor.ps1:
```diff
@@ -0,0 +1,102 @@
+            try {
+                $replServer = Get-DbaReplServer -SqlInstance $instance -SqlCredential $SqlCredential -EnableException:$EnableException
+            } catch {
+                Stop-Function -Message "Failure" -Category ConnectionError -ErrorRecord $_ -Target $instance -Continue
```
The "Faiulre" is only used for Connect calls so this should be updated to be more clear. If you want the catch block to actually hit you need to always pass in EnableException and not depend on the user. If you want to depend on the user, then remove the try/catch block completely.

wsmelton commented on public/Disable-DbaReplDistributor.ps1:
```diff
@@ -0,0 +1,102 @@
+                        $replServer.UninstallDistributor($Force)
+                    }
+                } catch {
+                    Stop-Function -Message "Unable to disable replication distribution" -ErrorRecord $_ -Target $instance -Continue
```
Update to "uninstall" if we move to change to the verb of the function

wsmelton commented on public/Disable-DbaReplDistributor.ps1:
```diff
@@ -0,0 +1,102 @@
+                $replServer
+
+            } else {
+                Stop-Function -Message "$instance isn't currently enabled for distributing." -ErrorRecord $_ -Target $instance -Continue
```
What error would get captured here in `$_`? How is this message reached if I use `-Force`?

wsmelton commented on public/Disable-DbaReplDistributor.ps1:
```diff
@@ -0,0 +1,102 @@
+            }
+            Write-Message -Level Verbose -Message "Disabling replication distribution for $instance"
+
+            if ($replServer.IsDistributor) {
```
What if I use `-Force` parameter?

wsmelton commented on public/Disable-DbaReplPublishing.ps1:
```diff
@@ -0,0 +1,100 @@
+function Disable-DbaReplPublishing {
```
Purely because of the method used and referenced this should get named as `Remove-DbaReplPublisher`

wsmelton replied:
Consider it from this thought: If I "disable" it, is the "enable" command going to just bring it back up or create it a new object? If enabling creates a new object that is not disable/enable then.

wsmelton replied:
Maybe we should use `Register` and `Unregister` instead?

wsmelton commented on public/Disable-DbaReplPublishing.ps1:
```diff
@@ -0,0 +1,100 @@
+function Disable-DbaReplPublishing {
+    <#
+    .SYNOPSIS
+        Disables publishing for the target SQL instances.
```
```suggestion
        Removing the publisher on the specific SQL Server instance(s).
```

wsmelton commented on public/Disable-DbaReplPublishing.ps1:
```diff
@@ -0,0 +1,100 @@
+        Disables publishing for the target SQL instances.
+
+    .DESCRIPTION
+        Disables publishing for the target SQL instances.
```
```suggestion
        Removes the registration for the Publisher from the Distributor on the specific SQL Server instance(s).
        This command is equivalent to running sp_dropdistpublisher stored procedure.
```

wsmelton commented on public/Disable-DbaReplPublishing.ps1:
```diff
@@ -0,0 +1,100 @@
+        if the Publisher is on a remote server that cannot be reached.
+
+    #>
+    [CmdletBinding(DefaultParameterSetName = "Default", SupportsShouldProcess, ConfirmImpact = 'High')]
```
```suggestion
    [CmdletBinding(SupportsShouldProcess, ConfirmImpact = 'High')]
```

wsmelton commented on public/Disable-DbaReplPublishing.ps1:
```diff
@@ -0,0 +1,100 @@
+    #>
+    [CmdletBinding(DefaultParameterSetName = "Default", SupportsShouldProcess, ConfirmImpact = 'High')]
+    param (
+        [parameter(Mandatory, ValueFromPipeline)]
```
```suggestion
        [Parameter(Mandatory, ValueFromPipeline)]
```

wsmelton commented on public/Disable-DbaReplPublishing.ps1:
```diff
@@ -0,0 +1,100 @@
+            try {
+                $replServer = Get-DbaReplServer -SqlInstance $instance -SqlCredential $SqlCredential
+            } catch {
+                Stop-Function -Message "Failure" -Category ConnectionError -ErrorRecord $_ -Target $instance -Continue
```
The "Faiulre" is only used for Connect calls so this should be updated to be more clear. If you want the catch block to actually hit you need to always pass in EnableException and not depend on the user. If you want to depend on the user, then remove the try/catch block completely.

wsmelton commented on public/Disable-DbaReplPublishing.ps1:
```diff
@@ -0,0 +1,100 @@
+    process {
+        foreach ($instance in $SqlInstance) {
+            try {
+                $replServer = Get-DbaReplServer -SqlInstance $instance -SqlCredential $SqlCredential
```
If you want the catch block hit then:

```suggestion
                $replServer = Get-DbaReplServer -SqlInstance $instance -SqlCredential $SqlCredential -EnableException
```

Otherwise, remove the try/catch block completely and just do
```
                $replServer = Get-DbaReplServer -SqlInstance $instance -SqlCredential $SqlCredential -EnableException
```

wsmelton commented on public/Disable-DbaReplPublishing.ps1:
```diff
@@ -0,0 +1,100 @@
+            }
+            Write-Message -Level Verbose -Message "Disabling publishing for $instance"
+
+            if ($replServer.IsPublisher) {
```
What happens if I pass in `-Force` should it be allowed passed this check?

wsmelton commented on public/Disable-DbaReplPublishing.ps1:
```diff
@@ -0,0 +1,100 @@
+            if ($replServer.IsPublisher) {
+                try {
+                    if ($PSCmdlet.ShouldProcess($instance, "Disabling publishing on $instance")) {
+                        # uninstall distribution
```
Assuming this should state `# removing Publisher`?

wsmelton commented on public/Disable-DbaReplPublishing.ps1:
```diff
@@ -0,0 +1,100 @@
+                        $replServer.DistributionPublishers.Remove($Force)
+                    }
+                } catch {
+                    Stop-Function -Message "Unable to disable replication publishing" -ErrorRecord $_ -Target $instance -Continue
```
Maybe?
```suggestion
                    Stop-Function -Message "Unable to remove the Publisher registration" -ErrorRecord $_ -Target $instance -Continue
```

wsmelton commented on public/Disable-DbaReplPublishing.ps1:
```diff
@@ -0,0 +1,100 @@
+                $replServer
+
+            } else {
+                Stop-Function -Message "$instance isn't currently enabled for publishing." -Continue -ContinueLabel main -Target $instance -Category InvalidData
```
Is it really InvalidData?

wsmelton commented on public/Disable-DbaReplPublishing.ps1:
```diff
@@ -0,0 +1,100 @@
+
+    .PARAMETER Force
+        A Boolean value that specifies whether the Publisher is uninstalled from the Distributor without verifying that
+        Publisher has also uninstalled the Distributor, if the Publisher is on a separate server.
```
Please remove the line break

wsmelton commented on public/Disable-DbaReplPublishing.ps1:
```diff
@@ -0,0 +1,100 @@
+        A Boolean value that specifies whether the Publisher is uninstalled from the Distributor without verifying that
+        Publisher has also uninstalled the Distributor, if the Publisher is on a separate server.
+
+        If true, all the replication objects associated with the Publisher are dropped even if the Publisher is on a remote server
```
Please remove the line break

wsmelton commented on public/Disable-DbaReplPublishing.ps1:
```diff
@@ -0,0 +1,100 @@
+        For MFA support, please use Connect-DbaInstance.
+
+    .PARAMETER Force
+        A Boolean value that specifies whether the Publisher is uninstalled from the Distributor without verifying that
```
```suggestion
        Specifies whether the Publisher is uninstalled from the Distributor without verifying that...
```

wsmelton commented on public/Disable-DbaReplPublishing.ps1:
```diff
@@ -0,0 +1,100 @@
+
+            if ($replServer.IsPublisher) {
+                try {
+                    if ($PSCmdlet.ShouldProcess($instance, "Disabling publishing on $instance")) {
```
```suggestion
                    if ($PSCmdlet.ShouldProcess($instance, "Removing the Publisher on $instance")) {
```

wsmelton commented on public/Disable-DbaReplPublishing.ps1:
```diff
@@ -0,0 +1,100 @@
+        Disables replication distribution for the mssql1 and mssql2 instances using a sql login.
+
+        Specifies force so all the replication objects associated with the Publisher are dropped even
+        if the Publisher is on a remote server that cannot be reached.
```
Please remove line breaks.

Update to `Remove` if we are using that verb.

wsmelton commented on public/Enable-DbaReplDistributor.ps1:
```diff
@@ -0,0 +1,92 @@
+        Enables distribution for the mssql1 instance and names the distribution database repDatabase.
+
+    #>
+    [CmdletBinding(DefaultParameterSetName = "Default", SupportsShouldProcess, ConfirmImpact = 'Medium')]
```
```suggestion
    [CmdletBinding(SupportsShouldProcess, ConfirmImpact = 'Medium')]
```

wsmelton commented on public/Enable-DbaReplDistributor.ps1:
```diff
@@ -0,0 +1,92 @@
+    #>
+    [CmdletBinding(DefaultParameterSetName = "Default", SupportsShouldProcess, ConfirmImpact = 'Medium')]
+    param (
+        [parameter(Mandatory, ValueFromPipeline)]
```
```suggestion
        [Parameter(Mandatory, ValueFromPipeline)]
```

wsmelton commented on public/Enable-DbaReplDistributor.ps1:
```diff
@@ -0,0 +1,92 @@
+            try {
+                $replServer = Get-DbaReplServer -SqlInstance $instance -SqlCredential $SqlCredential
+            } catch {
+                Stop-Function -Message "Failure" -Category ConnectionError -ErrorRecord $_ -Target $instance -Continue
```
This message is reserved for use with our connection command. As-is this catch will never be hit unless the Get call throws a hard exception.

If you want the catch block hit, then Get call should include -EnableException (not based on parameter being passed in). Otherwise the try/catch should be removed.

wsmelton commented on public/Enable-DbaReplDistributor.ps1:
```diff
@@ -0,0 +1,92 @@
+            }
+
+            $replServer.Refresh()
+            $replServer
```
I would not execpt this to be returned if it was not enabled?

jpomfret replied:
agreed - moved inside try block

wsmelton commented on public/Disable-DbaReplPublishing.ps1:
```diff
@@ -0,0 +1,100 @@
+                }
+
+                $replServer.Refresh()
+                $replServer
```
I would not expect this to be returned if it was not disabled?

jpomfret replied:
yeah agreed - moved up to inside the try block

wsmelton commented on public/Enable-DbaReplPublishing.ps1:
```diff
@@ -0,0 +1,118 @@
+        Attempts to set the procedure '[dbo].[StartUpProc1]' in the master database of SqlBox1\Instance2 for automatic execution when the instance is started.
+
+    #>
+    [CmdletBinding(DefaultParameterSetName = "Default", SupportsShouldProcess, ConfirmImpact = 'Medium')]
```
```suggestion
    [CmdletBinding(SupportsShouldProcess, ConfirmImpact = 'Medium')]
```

wsmelton commented on public/Enable-DbaReplPublishing.ps1:
```diff
@@ -0,0 +1,118 @@
+    #>
+    [CmdletBinding(DefaultParameterSetName = "Default", SupportsShouldProcess, ConfirmImpact = 'Medium')]
+    param (
+        [parameter(Mandatory, ValueFromPipeline)]
```
```suggestion
        [Parameter(Mandatory, ValueFromPipeline)]
```

wsmelton commented on public/Enable-DbaReplPublishing.ps1:
```diff
@@ -0,0 +1,118 @@
+            try {
+                if ($PSCmdlet.ShouldProcess($instance, "Enabling publishing on $instance")) {
+
+                    if ($replServer.IsDistributor) {
```
This should be outside the try/catch block. WhatIf output would be false if the distributor is not enabled.

wsmelton commented on public/Enable-DbaReplPublishing.ps1:
```diff
@@ -0,0 +1,118 @@
+            try {
+                $replServer = Get-DbaReplServer -SqlInstance $instance -SqlCredential $SqlCredential
+            } catch {
+                Stop-Function -Message "Failure" -Category ConnectionError -ErrorRecord $_ -Target $instance -Continue
```
Same as others, "Failure" is reserved for Connect commmand calls. If try/catch is desired then `-EnableException` should be added, otherwise remove the try/catch and let `Get-DbaReplServer` handle throwing errors via `-EnableException:$EnableException`

wsmelton commented on public/Enable-DbaReplPublishing.ps1:
```diff
@@ -0,0 +1,118 @@
+                        $distPublisher = New-Object Microsoft.SqlServer.Replication.DistributionPublisher
+                        $distPublisher.ConnectionContext = $replServer.ConnectionContext
+                        $distPublisher.Name = $instance #- name of the Publisher.
+                        $distPublisher.DistributionDatabase = $replServer.DistributionDatabases.Name #- the name of the database created in step 5.
```
What step 5?

jpomfret replied:
copy \ paste from the docs 🤣 removed comment!

wsmelton commented on public/Enable-DbaReplPublishing.ps1:
```diff
@@ -0,0 +1,118 @@
+                        $distPublisher.Name = $instance #- name of the Publisher.
+                        $distPublisher.DistributionDatabase = $replServer.DistributionDatabases.Name #- the name of the database created in step 5.
+
+                        if (-not $PSBoundParameters.SnapshotShare) {
```
```suggestion
                        if (Test-Bound SnapshotShare -Not) {
```

wsmelton commented on public/Get-DbaReplServer.ps1:
```diff
@@ -60,9 +60,14 @@ function Get-DbaRepServer {
-                # use System.Data instead of Microsoft.Data
-                $sqlconn = New-SqlConnection -SqlInstance $instance -SqlCredential $SqlCredential
-                New-Object Microsoft.SqlServer.Replication.ReplicationServer $sqlconn
+                $server = Connect-DbaInstance -SqlInstance $instance -SqlCredential $SqlCredential
```
We need to be consistent here if we are not going to use `New-SqlConnection` anymore then it should be removed from ` Test-DbaReplLatency` as well, completely.

We only created that command to deal with RMO initially I recall. So if we don't require it anymore, get rid of it.

jpomfret replied:
I've opened and Issue for `Test-DbaReplLatency` that doesn't work at all for me and so I've added on that issue that it should be rewritten to not use `New-SqlConnection`. I'll try and get that picked up once this stuff is complete.

https://github.com/dataplat/dbatools/issues/9028

wsmelton reviewed (commented):
I know this isn't included in this PR, but it plays a big role in the usage. 

`Add-ReplicationLibrary` being used as a private function for Get-DbaReplServer and particularly in testing has to remove the use of `EnableException`.

https://github.com/dataplat/dbatools/blob/50becdebf6271ca23f1ed3a4cf5d245e0b2519a5/private/functions/Add-ReplicationLibrary.ps1#L1-L15

This should not depend on `-EnableException` and should throw an error. `Stop-Function` should be passing out any error record returned by the `Add-Type` command. Very import, remove the `return` because we have seen that cause problems with a return is buried in private functions before.

I would also advise checking if the library isn't already loaded, maybe compare versions to what we package in the module so we can throw out more detail on failures...make it easier to support if they are using it on a server with SQL installed by at least making a warning message always appear.

wsmelton commented on public/Get-DbaReplServer.ps1:
```diff
@@ -60,9 +60,14 @@ function Get-DbaRepServer {
+                $replServer | Add-Member -Type NoteProperty -Name InstanceName -Value $server.ServiceName -Force
+                $replServer | Add-Member -Type NoteProperty -Name SqlInstance -Value $server.DomainInstanceName -Force
+
+                Select-DefaultView -InputObject $replServer -Property ComputerName, InstanceName, SqlInstance, DistributorInstalled, DistributorAvailable, IsDistributor, IsPublisher, HasRemotePublisher, DistributionServer, DistributionDatabase, WorkingDirectory, AgentCheckupInterval, DistributionDatabases, DistributionPublishers, ReplicationDatabases, RegisteredSubscribers
```
Not sure if there is possibility to lessen number of properties? Once you get over 7 (think) PowerShell converts it to a list. Consider what is bare minimum they need to know if they are running this interactively? I'd say: ComputerName, InstanceName, SqlInstance, IsDistributor, IsPublisher, DistributionServer, DistributionDatabase is sufficient for a "server" object return.

jpomfret replied:
Makes sense to me..
(Also the ReplicationDatabases include all the databases on the instance so is a bit confusing anyway tbh)

wsmelton commented on public/Get-DbaReplServer.ps1:
```diff
@@ -60,9 +60,14 @@ function Get-DbaRepServer {
+
+                Select-DefaultView -InputObject $replServer -Property ComputerName, InstanceName, SqlInstance, DistributorInstalled, DistributorAvailable, IsDistributor, IsPublisher, HasRemotePublisher, DistributionServer, DistributionDatabase, WorkingDirectory, AgentCheckupInterval, DistributionDatabases, DistributionPublishers, ReplicationDatabases, RegisteredSubscribers
             } catch {
                 Stop-Function -Message "Error occurred while establishing connection to $instance" -Category ConnectionError -ErrorRecord $_ -Target $instance -Continue
```
This should have our standard try/catch for the `$server` object. There is no reason to deviate from that for this command.

```powershell
try {
    $server = Connect-DbaInstance -SqlInstance $instance -SqlCredential $SqlCredential
} catch {
    Stop-Function -Message "Failure" -Category ConnectionError -ErrorRecord $_ -Target $instance -Continue
}```

wsmelton commented on public/Get-DbaReplDistributor.ps1:
```diff
@@ -59,12 +57,11 @@ function Get-DbaRepDistributor {
             # Connect to the distributor of the instance
             try {
                 $server = Connect-DbaInstance -SqlInstance $instance -SqlCredential $SqlCredential
```
This should be a separate try/catch and use our standard Failure message

wsmelton replied:
This also seems redudant to call the Connection command. Why do we need to make two separate connections for the same object to be output?

The ComputerName, ServiceName and DomainInstanceName is already on the distributor object because `Get-DBaReplServer` returns those properties.

jpomfret reviewed (commented):
Thanks @wsmelton for the detailed review! 

Apart from the question around renaming the enable\disable to new\remove I *think* I've resolved all issues and answered questions - let me know if I missed any

wsmelton commented on public/Get-DbaReplSubscription.ps1:
```diff
@@ -51,34 +68,52 @@ function Get-DbaReplSubscription {
         [PSCredential]$SqlCredential,
         [object[]]$Database,
-        [String]$Name,
+        [String[]]$Name,
```
```suggestion
        [string[]]$Name,
```

jpomfret commented:
@wsmelton \ @potatoqualitee 

Hey folks,
I've been chipping away at this and I'm not happy with how the subscription commands are coming together - for me to get a list of subscriptions I have to go to the publisher, and then get subscriptions from there - so because of this the `-SqlInstance` parameter on the `Get` command is the publisher, mssql1, in this example - however then for the `New` and `Remove` commands I have the subscriber SQL instance as the `-SqlInstance` parameter so mssql2 in this example - it feels confusing.

Talking to replication friends it sounds like replication is mostly looked at from the angle of  the publisher, so I think I'd like to swap these around - so instead of `-SqlInstance` being the subscription side it would be the publisher and then I'd change `-PublisherSqlInstance` to be `-SubscriberSqlInstance`.

What do you think? Any massive objections? Just feels clunky the current implementation. 

```PowerShell
$sub = @{
    SqlInstance               = 'mssql2'
    Database                  = 'ReplDb'
    PublicationDatabase       = 'ReplDb'
    PublisherSqlInstance      = 'mssql1'
    PublicationName           = 'testtrans'
    Type                      = 'Push'
}
New-DbaReplSubscription @sub

$sub = @{
    SqlInstance          = 'mssql1' # have to target publisher
    Database             = 'ReplDb'
    SubscriptionDatabase = 'ReplDb'
    Name                 = 'testPubs'
    Type                 = 'Push' 
}
Get-DbaReplSubscription @sub

$sub = @{
    SqlInstance          = 'mssql2' 
    SubscriptionDatabase = 'pubs'
    PublisherSqlInstance = 'mssql1'
    PublicationDatabase  = 'pubs'
    PublicationName      = 'testPub'
}
Remove-DbaReplSubscription @sub
```

jpomfret commented:
Tests are now passing and so are all the new gh workflow ones specific for replication :) 

@wsmelton - how we looking here? Any others things we need to do?

jpomfret commented:
Hi @wsmelton,

I built a little lab in Azure with AutomatedLab to test replication out in a windows\domain environment. I ran through:
- enabling distribution
- enabling publishing
- create 3 publications (1 transactional, 1 merge, 1 snapshot)
- add an article to each
- add subscription to each
- ensure replication was working at this point
- remove subscriptions
- remove articles
- remove publications
- disable publishing
- disable distribution

I fixed a couple of issues, related to using other schemas, which I've now fixed and also found one issue with `Disable-DbaReplPublishing` that I will look into:
```
<#
WARNING: [12:32:22][Disable-DbaReplPublishing] Unable to disable replication publishing | Cannot drop server 'SQL2017'
as Distribution Publisher because there are databases enabled for replication on that server.
Changed database context to 'distribution'.
#>
```

Transcript of testing:
```
**********************
Windows PowerShell transcript start
Start time: 20230809122354
Username: dbatools\a405201
RunAs User: dbatools\a405201
Machine: POSHClient1 (Microsoft Windows NT 10.0.14393.0)
Host Application: C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe
Process ID: 976
PSVersion: 5.1.14393.5582
PSEdition: Desktop
PSCompatibleVersions: 1.0, 2.0, 3.0, 4.0, 5.0, 5.1.14393.5582
BuildVersion: 10.0.14393.5582
CLRVersion: 4.0.30319.42000
WSManStackVersion: 3.0
PSRemotingProtocolVersion: 2.3
SerializationVersion: 1.1.0.1
**********************
Transcript started, output file is test_replication.txt
PS C:\github\jpomfret\dbatools> import-module .\dbatools.psd1 -force
PS C:\github\jpomfret\dbatools> Set-DbatoolsConfig -FullName sql.connection.encrypt -Value optional
PS C:\github\jpomfret\dbatools> Set-DbatoolsConfig -FullName sql.connection.trustcert -Value $true
PS C:\github\jpomfret\dbatools> # Get the distributor
PS C:\github\jpomfret\dbatools> Get-DbaReplDistributor -SqlInstance sql2017


ComputerName         : SQL2017
InstanceName         : MSSQLSERVER
SqlInstance          : SQL2017
IsPublisher          : False
IsDistributor        : False
DistributionServer   :
DistributionDatabase :
DistributorInstalled : False
DistributorAvailable : False
HasRemotePublisher   : False



PS C:\github\jpomfret\dbatools> # view publications
PS C:\github\jpomfret\dbatools> Get-DbaReplPublication -SqlInstance sql2017
PS C:\github\jpomfret\dbatools> # view articles
PS C:\github\jpomfret\dbatools> Get-DbaReplArticle -SqlInstance sql2017
PS C:\github\jpomfret\dbatools> # get subscriptions
PS C:\github\jpomfret\dbatools> Get-DbaReplSubscription -SqlInstance sql2017
PS C:\github\jpomfret\dbatools> # enable distribution
PS C:\github\jpomfret\dbatools> Enable-DbaReplDistributor -SqlInstance sql2017


ComputerName         : SQL2017
InstanceName         : MSSQLSERVER
SqlInstance          : SQL2017
IsDistributor        : True
IsPublisher          : False
DistributionServer   : SQL2017
DistributionDatabase :



PS C:\github\jpomfret\dbatools> # enable publishing
PS C:\github\jpomfret\dbatools> Enable-DbaReplPublishing -SqlInstance sql2017


ComputerName         : SQL2017
InstanceName         : MSSQLSERVER
SqlInstance          : SQL2017
IsDistributor        : True
IsPublisher          : True
DistributionServer   : SQL2017
DistributionDatabase : distribution



PS C:\github\jpomfret\dbatools> # Get the distributor
PS C:\github\jpomfret\dbatools> Get-DbaReplDistributor -SqlInstance sql2017


ComputerName         : SQL2017
InstanceName         : MSSQLSERVER
SqlInstance          : SQL2017
IsPublisher          : True
IsDistributor        : True
DistributionServer   : SQL2017
DistributionDatabase : distribution
DistributorInstalled : True
DistributorAvailable : True
HasRemotePublisher   : False



PS C:\github\jpomfret\dbatools> # add a transactional publication
PS C:\github\jpomfret\dbatools> $pub = @{
    SqlInstance = 'sql2017'
    Database    = 'AdventureWorksLT2017'
    Name        = 'testPub'
    Type        = 'Transactional'
}
PS C:\github\jpomfret\dbatools> New-DbaReplPublication @pub


ComputerName  : SQL2017
InstanceName  : MSSQLSERVER
SQLInstance   : [sql2017]
DatabaseName  : AdventureWorksLT2017
Name          : testPub
Type          : Transactional
Articles      : {}
Subscriptions : {}



PS C:\github\jpomfret\dbatools> # add a merge publication
PS C:\github\jpomfret\dbatools> $pub = @{
    SqlInstance = 'sql2017'
    Database    = 'AdventureWorksLT2017'
    Name        = 'mergey'
    Type        = 'Merge'
}
PS C:\github\jpomfret\dbatools> New-DbaReplPublication @pub


ComputerName  : SQL2017
InstanceName  : MSSQLSERVER
SQLInstance   : [sql2017]
DatabaseName  : AdventureWorksLT2017
Name          : mergey
Type          : Merge
Articles      : {}
Subscriptions : {}



PS C:\github\jpomfret\dbatools> # add a snapshot publication
PS C:\github\jpomfret\dbatools> $pub = @{
    SqlInstance = 'sql2017'
    Database    = 'AdventureWorksLT2017'
    Name        = 'snappy'
    Type        = 'Snapshot'
}
PS C:\github\jpomfret\dbatools> New-DbaReplPublication @pub


ComputerName  : SQL2017
InstanceName  : MSSQLSERVER
SQLInstance   : [sql2017]
DatabaseName  : AdventureWorksLT2017
Name          : snappy
Type          : Snapshot
Articles      : {}
Subscriptions : {}



PS C:\github\jpomfret\dbatools> # view publications
PS C:\github\jpomfret\dbatools> Get-DbaReplPublication -SqlInstance sql2017


ComputerName  : SQL2017
InstanceName  : MSSQLSERVER
SQLInstance   : [sql2017]
DatabaseName  : AdventureWorksLT2017
Name          : snappy
Type          : Snapshot
Articles      : {}
Subscriptions : {}

ComputerName  : SQL2017
InstanceName  : MSSQLSERVER
SQLInstance   : [sql2017]
DatabaseName  : AdventureWorksLT2017
Name          : testPub
Type          : Transactional
Articles      : {}
Subscriptions : {}

ComputerName  : SQL2017
InstanceName  : MSSQLSERVER
SQLInstance   : [sql2017]
DatabaseName  : AdventureWorksLT2017
Name          : mergey
Type          : Merge
Articles      : {}
Subscriptions : {}



PS C:\github\jpomfret\dbatools> # add an article to each publication
PS C:\github\jpomfret\dbatools> $article = @{
    SqlInstance = 'sql2017'
    Database    = 'AdventureWorksLT2017'
    Publication = 'testpub'
    Schema      = 'salesLT'
    Name        = 'customer'
    Filter      = "lastname = 'gates'"
}
PS C:\github\jpomfret\dbatools> Add-DbaReplArticle @article


ComputerName      : SQL2017
InstanceName      : MSSQLSERVER
SqlInstance       : [sql2017]
DatabaseName      : AdventureWorksLT2017
PublicationName   : testPub
Name              : customer
Type              : LogBased
VerticalPartition : False
SourceObjectOwner : SalesLT
SourceObjectName  : Customer



PS C:\github\jpomfret\dbatools> $article = @{
    SqlInstance = 'sql2017'
    Database    = 'AdventureWorksLT2017'
    Publication = 'Mergey'
    Schema      = 'salesLT'
    Name        = 'product'
}
PS C:\github\jpomfret\dbatools> Add-DbaReplArticle @article


ComputerName      : SQL2017
InstanceName      : MSSQLSERVER
SqlInstance       : [sql2017]
DatabaseName      : AdventureWorksLT2017
PublicationName   : mergey
Name              : product
Type              : TableBased
VerticalPartition : False
SourceObjectOwner : SalesLT
SourceObjectName  : Product



PS C:\github\jpomfret\dbatools> $article = @{
    SqlInstance = 'sql2017'
    Database    = 'AdventureWorksLT2017'
    Publication = 'snappy'
    Schema      = 'salesLT'
    Name        = 'address'
}
PS C:\github\jpomfret\dbatools> Add-DbaReplArticle @article


ComputerName      : SQL2017
InstanceName      : MSSQLSERVER
SqlInstance       : [sql2017]
DatabaseName      : AdventureWorksLT2017
PublicationName   : snappy
Name              : address
Type              : LogBased
VerticalPartition : False
SourceObjectOwner : SalesLT
SourceObjectName  : Address



PS C:\github\jpomfret\dbatools> # view articles
PS C:\github\jpomfret\dbatools> Get-DbaReplArticle -SqlInstance sql2017


ComputerName      : SQL2017
InstanceName      : MSSQLSERVER
SqlInstance       : [sql2017]
DatabaseName      : AdventureWorksLT2017
PublicationName   : snappy
Name              : address
Type              : LogBased
VerticalPartition : False
SourceObjectOwner : SalesLT
SourceObjectName  : Address

ComputerName      : SQL2017
InstanceName      : MSSQLSERVER
SqlInstance       : [sql2017]
DatabaseName      : AdventureWorksLT2017
PublicationName   : testPub
Name              : customer
Type              : LogBased
VerticalPartition : False
SourceObjectOwner : SalesLT
SourceObjectName  : Customer

ComputerName      : SQL2017
InstanceName      : MSSQLSERVER
SqlInstance       : [sql2017]
DatabaseName      : AdventureWorksLT2017
PublicationName   : mergey
Name              : product
Type              : TableBased
VerticalPartition : False
SourceObjectOwner : SalesLT
SourceObjectName  : Product



PS C:\github\jpomfret\dbatools> # add subscriptions
PS C:\github\jpomfret\dbatools> $sub = @{
    SqlInstance           = 'sql2017'
    Database              = 'AdventureWorksLT2017'
    SubscriberSqlInstance = 'sql2019'
    SubscriptionDatabase  = 'AdventureWorksLT2017'
    PublicationName       = 'testpub'
    Type                  = 'Push'
}
PS C:\github\jpomfret\dbatools> New-DbaReplSubscription @sub
PS C:\github\jpomfret\dbatools> $sub = @{
    SqlInstance           = 'sql2017'
    Database              = 'AdventureWorksLT2017'
    SubscriberSqlInstance = 'sql2019'
    SubscriptionDatabase  = 'AdventureWorksLT2017Merge'
    PublicationName       = 'Mergey'
    Type                  = 'Push'
}
PS C:\github\jpomfret\dbatools> New-DbaReplSubscription @sub
PS C:\github\jpomfret\dbatools> $sub = @{
    SqlInstance           = 'sql2017'
    Database              = 'AdventureWorksLT2017'
    SubscriberSqlInstance = 'sql2019'
    SubscriptionDatabase  = 'AdventureWorksLT2017Snap'
    PublicationName       = 'Snappy'
    Type                  = 'Push'
}
PS C:\github\jpomfret\dbatools> New-DbaReplSubscription @sub
PS C:\github\jpomfret\dbatools> # start snapshot agent
PS C:\github\jpomfret\dbatools> Get-DbaAgentJob -SqlInstance sql2017 -Category repl-snapshot | Start-DbaAgentJob


ComputerName           : SQL2017
InstanceName           : MSSQLSERVER
SqlInstance            : SQL2017
Name                   : SQL2017-AdventureWorksLT2017-mergey-2
Category               : REPL-Snapshot
OwnerLoginName         : dbatools\a405201
CurrentRunStatus       : Executing
CurrentRunRetryAttempt : 0
Enabled                : True
LastRunDate            : 01/01/0001 00:00:00
LastRunOutcome         : Unknown
HasSchedule            : True
OperatorToEmail        :
CreateDate             : 09/08/2023 12:25:05

ComputerName           : SQL2017
InstanceName           : MSSQLSERVER
SqlInstance            : SQL2017
Name                   : SQL2017-AdventureWorksLT2017-snappy-3
Category               : REPL-Snapshot
OwnerLoginName         : dbatools\a405201
CurrentRunStatus       : Executing
CurrentRunRetryAttempt : 0
Enabled                : True
LastRunDate            : 01/01/0001 00:00:00
LastRunOutcome         : Unknown
HasSchedule            : True
OperatorToEmail        :
CreateDate             : 09/08/2023 12:25:10

ComputerName           : SQL2017
InstanceName           : MSSQLSERVER
SqlInstance            : SQL2017
Name                   : SQL2017-AdventureWorksLT2017-testPub-1
Category               : REPL-Snapshot
OwnerLoginName         : dbatools\a405201
CurrentRunStatus       : Executing
CurrentRunRetryAttempt : 0
Enabled                : True
LastRunDate            : 01/01/0001 00:00:00
LastRunOutcome         : Unknown
HasSchedule            : True
OperatorToEmail        :
CreateDate             : 09/08/2023 12:25:02



PS C:\github\jpomfret\dbatools> # remove subscriptions
PS C:\github\jpomfret\dbatools> $sub = @{
    SqlInstance           = 'sql2017'
    Database              = 'AdventureWorksLT2017'
    SubscriptionDatabase  = 'AdventureWorksLT2017'
    SubscriberSqlInstance = 'sql2019'
    PublicationName       = 'testPub'
}
PS C:\github\jpomfret\dbatools> Remove-DbaReplSubscription @sub
PS C:\github\jpomfret\dbatools> $sub = @{
    SqlInstance           = 'sql2017'
    Database              = 'AdventureWorksLT2017'
    SubscriptionDatabase  = 'AdventureWorksLT2017'
    SubscriberSqlInstance = 'sql2019'
    PublicationName       = 'Mergey'
}
PS C:\github\jpomfret\dbatools> Remove-DbaReplSubscription @sub
WARNING: Didn't find a subscription to Mergey on sql2019.AdventureWorksLT2017
PS C:\github\jpomfret\dbatools> $sub = @{
    SqlInstance           = 'sql2017'
    Database              = 'AdventureWorksLT2017'
    SubscriptionDatabase  = 'AdventureWorksLT2017Merge'
    SubscriberSqlInstance = 'sql2019'
    PublicationName       = 'Mergey'
}
Remove-DbaReplSubscription @sub
PS C:\github\jpomfret\dbatools> $sub = @{
    SqlInstance           = 'sql2017'
    Database              = 'AdventureWorksLT2017'
    SubscriptionDatabase  = 'AdventureWorksLT2017Snap'
    SubscriberSqlInstance = 'sql2019'
    PublicationName       = 'snappy'
}
Remove-DbaReplSubscription @sub
PS C:\github\jpomfret\dbatools> # remove an article
$article = @{
    SqlInstance = 'sql2017'
    Database    = 'AdventureWorksLT2017'
    Publication = 'testpub'
    Schema      = 'salesLT'
    Name        = 'customer'
}
Remove-DbaReplArticle @article

Confirm
Are you sure you want to perform this action?
Performing the operation "Removing the article SalesLT.Customer from the testPub publication on [sql2017]" on target "customer".
&Yes  Yes to &All  &No  No to A&ll  &Suspend
Y


ComputerName : SQL2017
InstanceName : MSSQLSERVER
SqlInstance  : [sql2017]
Database     : AdventureWorksLT2017
ObjectName   : Customer
ObjectSchema : SalesLT
Status       : Removed
IsRemoved    : True



PS C:\github\jpomfret\dbatools> # remove an article
$article = @{
    SqlInstance = 'sql2017'
    Database    = 'AdventureWorksLT2017'
    Publication = 'Mergey'
    Schema      = 'salesLT'
    Name        = 'product'
}
Remove-DbaReplArticle @article
Confirm
Are you sure you want to perform this action?
Performing the operation "Removing the article SalesLT.Product from the mergey publication on [sql2017]" on target "product".
&Yes  Yes to &All  &No  No to A&ll  &Suspend
Y


ComputerName : SQL2017
InstanceName : MSSQLSERVER
SqlInstance  : [sql2017]
Database     : AdventureWorksLT2017
ObjectName   : Product
ObjectSchema : SalesLT
Status       : Removed
IsRemoved    : True



PS C:\github\jpomfret\dbatools> # remove an article
$article = @{
    SqlInstance = 'sql2017'
    Database    = 'AdventureWorksLT2017'
    Publication = 'snappy'
    Schema      = 'salesLT'
    Name        = 'address'
}
Remove-DbaReplArticle @article
Confirm
Are you sure you want to perform this action?
Performing the operation "Removing the article SalesLT.Address from the snappy publication on [sql2017]" on target "address".
&Yes  Yes to &All  &No  No to A&ll  &Suspend
Y


ComputerName : SQL2017
InstanceName : MSSQLSERVER
SqlInstance  : [sql2017]
Database     : AdventureWorksLT2017
ObjectName   : Address
ObjectSchema : SalesLT
Status       : Removed
IsRemoved    : True



PS C:\github\jpomfret\dbatools> ## remove publications
$pub = @{
    SqlInstance = 'sql2017'
    Database    = 'AdventureWorksLT2017'
    Name        = 'TestPub'
}
Remove-DbaReplPublication @pub

Confirm
Are you sure you want to perform this action?
Performing the operation "Removing the publication testPub on [sql2017]" on target "testPub".
&Yes  Yes to &All  &No  No to A&ll  &Suspend
Y
Confirm
Are you sure you want to perform this action?
Performing the operation "Stopping the REPL-LogReader job for the database AdventureWorksLT2017 on [sql2017]" on target "testPub".
&Yes  Yes to &All  &No  No to A&ll  &Suspend
Y


ComputerName : SQL2017
InstanceName : MSSQLSERVER
SqlInstance  : [sql2017]
Database     : AdventureWorksLT2017
Name         : testPub
Type         : Transactional
Status       : Removed
IsRemoved    : True



PS C:\github\jpomfret\dbatools> $pub = @{
    SqlInstance = 'sql2017'
    Database    = 'AdventureWorksLT2017'
    Name        = 'Snappy'
}
Remove-DbaReplPublication @pub
Confirm
Are you sure you want to perform this action?
Performing the operation "Removing the publication snappy on [sql2017]" on target "snappy".
&Yes  Yes to &All  &No  No to A&ll  &Suspend
Y
Confirm
Are you sure you want to perform this action?
Performing the operation "Stopping the REPL-LogReader job for the database AdventureWorksLT2017 on [sql2017]" on target "snappy".
&Yes  Yes to &All  &No  No to A&ll  &Suspend
Y


ComputerName : SQL2017
InstanceName : MSSQLSERVER
SqlInstance  : [sql2017]
Database     : AdventureWorksLT2017
Name         : snappy
Type         : Snapshot
Status       : Removed
IsRemoved    : True



PS C:\github\jpomfret\dbatools> 
$pub = @{
    SqlInstance = 'sql2017'
    Database    = 'AdventureWorksLT2017'
    Name        = 'Mergey'
}
Remove-DbaReplPublication @pub

Confirm
Are you sure you want to perform this action?
Performing the operation "Removing the publication mergey on [sql2017]" on target "mergey".
&Yes  Yes to &All  &No  No to A&ll  &Suspend
Y
Confirm
Are you sure you want to perform this action?
Performing the operation "Stopping the REPL-LogReader job for the database AdventureWorksLT2017 on [sql2017]" on target "mergey".
&Yes  Yes to &All  &No  No to A&ll  &Suspend
Y


ComputerName : SQL2017
InstanceName : MSSQLSERVER
SqlInstance  : [sql2017]
Database     : AdventureWorksLT2017
Name         : mergey
Type         : Merge
Status       : Removed
IsRemoved    : True



PS C:\github\jpomfret\dbatools> # disable publishing 
Disable-DbaReplPublishing -SqlInstance sql2017 -force

Confirm
Are you sure you want to perform this action?
Performing the operation "Disabling and removing publishing on sql2017" on target "sql2017".
&Yes  Yes to &All  &No  No to A&ll  &Suspend
Y
WARNING: [12:32:22][Disable-DbaReplPublishing] Unable to disable replication publishing | Cannot drop server 'SQL2017' as Distribution Publisher because there are databases enabled for replication on that server.
Changed database context to 'distribution'.
PS C:\github\jpomfret\dbatools> TerminatingError(): "System error."
PS C:\github\jpomfret\dbatools> Disable-DbaReplDistributor -SqlInstance  sql2017
Confirm
Are you sure you want to perform this action?
Performing the operation "Disabling and removing distribution on sql2017" on target "sql2017".
&Yes  Yes to &All  &No  No to A&ll  &Suspend
Y


ComputerName         : SQL2017
InstanceName         : MSSQLSERVER
SqlInstance          : SQL2017
IsDistributor        : False
IsPublisher          : False
DistributionServer   : SQL2017
DistributionDatabase :


PS C:\github\jpomfret\dbatools> Get-DbaReplServer -SqlInstance sql2017


ComputerName         : SQL2017
InstanceName         : MSSQLSERVER
SqlInstance          : SQL2017
IsDistributor        : False
IsPublisher          : False
DistributionServer   :
DistributionDatabase :



PS C:\github\jpomfret\dbatools> Stop-Transcript
**********************
Windows PowerShell transcript end
End time: 20230809123302
**********************
```

wsmelton commented on dbatools.psd1:
```diff
@@ -693,7 +694,7 @@
         'Test-DbaPath',
         'Test-DbaPowerPlan',
-        'Test-DbaRepLatency',
+        'Test-DbaReplLatency',
```
I think we are safe to move this to the section below.

wsmelton commented on dbatools.psd1:
```diff
@@ -350,9 +350,10 @@
+        'Get-DbaReplDistributor',
+        'Get-DbaReplPublication',
+        'Get-DbaReplPublisher',
+        'Get-DbaReplServer',
```
Think we are safe to move these to the section below, keep them all organized together.

jpomfret commented on public/Enable-DbaReplPublishing.ps1:
```diff
@@ -0,0 +1,119 @@
+    process {
+        foreach ($instance in $SqlInstance) {
+
+            $replServer = Get-DbaReplServer -SqlInstance $instance -SqlCredential $SqlCredential
```
Have gone through and made sure ` -EnableException:$EnableException` is included on all public functions - thanks

jpomfret reviewed (commented):
Hey @wsmelton thanks for going through this - I've addressed all your concerns (I think) except the naming of these four commands:

- Enable-DbaReplDistributor
- Disable-DbaReplDistributor
- Enable-DbaReplPublishing
- Disable-DbaReplPublishing

Also I can only apologise for the massive PR - it was hard to break into chunks since so many of the commands work together.

jpomfret commented:
failed tests seem to be unrelated

potatoqualitee commented:
> failed tests seem to be unrelated

Login to appveyor using your GH creds and see if you can do the rerun incomplete button

potatoqualitee commented:
@wsmelton @jpomfret sorry for the delay. I can see it both ways. You Enable Replication but create a New Publisher eh?

What verbiage do you use when talking to a customer?

If it's not Enable/Disable, that would be the only thing I could see that would override SSMS.

potatoqualitee commented:
YOOOO lez do this! Congrats!! <3

potatoqualitee commented:
Thank you both so very much, what a huge amount of work. Looking forward to promoting this on socials 💯
